### ISSUE TITLE
[BUG] LSP plugin on vim gets crazy when dart-code-metrics is used

### ISSUE THREAD

**Issue by erickzanardo** [dart-code-checker/dart-code-metrics#262]
Hello,

We have been using `dart-code-metrics` on [Flame](https://github.com/flame-engine/flame) from some time and I have noticed that when the plugin is configured on the `analysis_options.yml`, the analyzer starts to get "crazy" and point really strange errors, like this one:

![dart-code-metrics](https://user-images.githubusercontent.com/835641/114237991-42b1ff80-995a-11eb-84af-99555718a5c3.png)

Notice the weird error on "Undefined class" seems like a "confused" code was sent to the analyzer.

I am not sure if this is a problem on dart-code-metrics, or in the vim plugin, but I thought on trying here first as everything works fine if I remove the dart-code-metrics from the plugin section.

Thanks for the time to read this issue and I am willing to help with whatever needed to solve this.

* **Dart code metrics version:** Tested both on `3.0.0-nullsafety.1` and `3.0.0-nullsafety.2`
* **Dart sdk version:** 2.12.2

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```yaml
dependencies:
  flutter:
    sdk: flutter
  meta: ^1.3.0
  ordered_set: ^3.0.0
  vector_math: '>=2.1.0 <3.0.0'

dev_dependencies:
  flutter_test:
    sdk: flutter
  test: ^1.16.0
  dart_code_metrics: ^3.0.0-nullsafety.2
  dartdoc: ^0.42.0

environment:
  sdk: ">=2.12.0 <3.0.0"
  flutter: ">=1.24.0"

flutter:
  uses-material-design: false
```

```yml
analyzer:
  strong-mode:
    implicit-casts: false
    implicit-dynamic: false

linter:
  rules:
    - always_declare_return_types
    - always_put_control_body_on_new_line
    - always_require_non_null_named_parameters
    - annotate_overrides
    - avoid_double_and_int_checks
    - avoid_dynamic_calls
    - avoid_empty_else
    - avoid_equals_and_hash_code_on_mutable_classes
    - avoid_escaping_inner_quotes
    - avoid_field_initializers_in_const_classes
    - avoid_init_to_null
    - avoid_js_rounded_ints
    - avoid_null_checks_in_equality_operators
    - avoid_private_typedef_functions
    - avoid_redundant_argument_values
    - avoid_relative_lib_imports
    - avoid_return_types_on_setters
    - avoid_shadowing_type_parameters
    - avoid_slow_async_io
    - avoid_type_to_string
    - avoid_types_as_parameter_names
    - avoid_unused_constructor_parameters
    - await_only_futures
    - camel_case_extensions
    - camel_case_types
    - cancel_subscriptions
    - cast_nullable_to_non_nullable
    - close_sinks
    - comment_references
    - constant_identifier_names
    - control_flow_in_finally
    - curly_braces_in_flow_control_structures
    - directives_ordering
    - do_not_use_environment
    - empty_catches
    - empty_constructor_bodies
    - empty_statements
    - exhaustive_cases
    - file_names
    - hash_and_equals
    - implementation_imports
    - invariant_booleans
    - iterable_contains_unrelated_type
    - join_return_with_assignment
    - library_names
    - library_prefixes
    - list_remove_unrelated_type
    - literal_only_boolean_expressions
    - missing_whitespace_between_adjacent_strings
    - no_adjacent_strings_in_list
    - no_duplicate_case_values
    - no_runtimeType_toString
    - omit_local_variable_types
    - package_api_docs
    - package_names
    - package_prefixed_library_names
    - parameter_assignments
    - prefer_adjacent_string_concatenation
    - prefer_asserts_in_initializer_lists
    - prefer_collection_literals
    - prefer_conditional_assignment
    - prefer_const_constructors
    - prefer_const_constructors_in_immutables
    - prefer_const_declarations
    - prefer_const_literals_to_create_immutables
    - prefer_contains
    - prefer_equal_for_default_values
    - prefer_final_fields
    - prefer_final_in_for_each
    - prefer_final_locals
    - prefer_for_elements_to_map_fromIterable
    - prefer_foreach
    - prefer_function_declarations_over_variables
    - prefer_generic_function_type_aliases
    - prefer_if_elements_to_conditional_expressions
    - prefer_if_null_operators
    - prefer_initializing_formals
    - prefer_inlined_adds
    - prefer_interpolation_to_compose_strings
    - prefer_is_empty
    - prefer_is_not_empty
    - prefer_is_not_operator
    - prefer_iterable_whereType
    - prefer_mixin
    - prefer_null_aware_operators
    - prefer_single_quotes
    - prefer_spread_collections
    - prefer_relative_imports
    - prefer_typing_uninitialized_variables
    - prefer_void_to_null
    - provide_deprecation_message
    - recursive_getters
    - slash_for_doc_comments
    - sort_unnamed_constructors_first
    - test_types_in_equals
    - throw_in_finally
    - type_annotate_public_apis
    - type_init_formals
    - unnecessary_await_in_return
    - unnecessary_brace_in_string_interps
    - unnecessary_const
    - unnecessary_getters_setters
    - unnecessary_lambdas
    - unnecessary_new
    - unnecessary_null_aware_assignments
    - unnecessary_null_in_if_null_operators
    - unnecessary_overrides
    - unnecessary_parenthesis
    - unnecessary_raw_strings
    - unnecessary_statements
    - unnecessary_string_escapes
    - unnecessary_string_interpolations
    - unnecessary_this
    - use_full_hex_values_for_flutter_colors
    - use_function_type_syntax_for_parameters
    - use_is_even_rather_than_modulo
    - use_rethrow_when_possible
    - unrelated_type_equality_checks
    - unsafe_html
    - void_checks

dart_code_metrics:
  rules:
    - prefer-trailing-comma
    - prefer-trailing-comma-for-collection
    - no-equal-then-else
    - no-object-declaration
    - potential-null-dereference
  metrics-exclude:
    - test/**
  metrics:
    number-of-parameters: 8
    number-of-methods: 32
    lines-of-executable-code: 200
    cyclomatic-complexity: 36
```
Both files can also be seen here on github: https://github.com/flame-engine/flame/tree/main/packages/flame

Not sure if this could help, but my nvim config is available here: https://github.com/erickzanardo/dotfiles/blob/master/init.vim
</details>

**Are you willing to submit a pull request to fix this bug?** Would need some guidance, but yes!

**Comment by christopherfujino** [comment]
I can reproduce this (with the latest flame repo) with the coc.nvim plugin configured as such:

```
  "dart": {
    "command": "dart",
    "args": [
      "/path/to/flutter/bin/cache/dart-sdk/bin/snapshots/analysis_server.dart.snapshot",
      "--lsp",
      "--client-id",
      "vim",
      "--client-version",
      "coc-nvim"
    ],
    "filetypes": ["dart"],
    "trace.server": "verbose"
  }
```

**Comment by luanpotter** [comment]
I can confirm I also saw the same issue on VSCode (on the flame repo), using the standard Dart and Flutter plugins. I thought it was something wrong with vscode at first and didn't bother much (a simple restart fixes it), but it has the same effect as it seems the underlying dart/flutter compiler is getting jumbled code to parse that does not match what is on the editor.

**Comment by incendial** [comment]
Hi! That looks very strange, firstly because we don't perform those checks that you shown, they are performed by the analyzer itself.

I've never used nvim and decided to run some checks in VS Code. Here is the result:
1. Before `3.0.0-nullsafety.2` we had a bug with the plugin bootstrapping and basically it was not even subscribed to file analysis at all. So that means that the plugin was not sending anything to the analysis server. 
2. I've encountered with a different issue in VS Code, but I think that they have the same problem underneath - if the `plugins` entry is added to the `analysis_options`, then the analyser starts working like that

https://user-images.githubusercontent.com/21116742/114265113-b35b2980-99f7-11eb-9643-517fa052b62c.mov

Or in other words - each time the file is updated and not saved, error disappear. And if I remove `plugins` and restart the analysis server then everything works fine and without this glowing. And again, I used `3.0.0-nullsafety.1` where plugin does nothing.

I don't think that the problem is with the VS Code extension or with the vim plugin. Seems like it's somewhere in the analysis server. Unless I'm missing something.

Have you used `dart_code_metrics` with LSP before 3.0.0? Have you faced the same problem or everything was fine?

@luanpotter do you have LSP preview setting enabled in VS Code or the problem remains without it?

**Comment by DanTup** [comment]
If anyone can repro the issue with server/client appearing to be out of sync in VS Code, I'd definitely be interested in steps.

You can also run the **Dart: Open Analyzer Diagnostics** command in the VS Code command palette to access the overlay info (click Context, then find the file and click the "Contents" link) - this shows what the server thinks the file looks like:

<img width="989" alt="Screenshot 2021-04-14 at 18 40 34" src="https://user-images.githubusercontent.com/1078012/114754832-eb909e00-9d50-11eb-94bc-4782dba08d97.png">

An instrumentation log file (captured with the analyzers `--instrumentation-log-file` flag, or the `dart.analyzerInstrumentationLogFile` setting in VS Code) might also be helpful.

Someone else reported "out of order edits" to me in VS Code before - they were also using a Vim extension (which sends the keystrokes through the extension host - I wonder if something async is causing them to get out of sync) but I was never able to repro or get a log from them.

**Comment by luanpotter** [comment]
@incendial sorry for the delay, it's been a busy week.

it's not trivial to test because it only gets "crazy" every so often, and every time it does I have to restart vscode. But here is some information:

* yes I do use the vim plugin for vscode
* I do have lsp preview enabled. with that setting on (which was on by default on my vscode) I am extremely confident, after a lot of testing, that it does happen every so often and when I remove the dart-code-metrics dependency it never happens. I tried this a lot so I am convinced having dart_code_metrics on the pubspec dev dependencies is one of the triggers.
* I tried a bit setting the preview enabled to false. it indeed seemed to fix it. I have been using it for a while like that and I have not encountered the error once. but I did not test it exhaustively so I cannot be as sure about this part

I will keep using with preview=off since I don't even know what it does and that for now seems to have fixed my issue. If it happens again in that scenario I will post back here. But eventually the preview might become the only option, so might be worth more investigating.

**Comment by incendial** [comment]
> But eventually the preview might become the only option, so might be worth more investigating.

@luanpotter I agree and that's why it will be very helpful to get a reproduction example for VS Code. Are there any chance that you can create one or share reproduction steps? I'm talking about the initial problem with "confused" code.

**Comment by luanpotter** [comment]
Sorry for the delay, but I can definitely and without a shadow of doubt now, due to extensive use, that setting `preview=off` fix the issue for me. It never happened again since :)

**Comment by DanTup** [comment]
FWIW, LSP is going to become the default (and then only) option in the near future. If there are issues in the LSP server it's important they're tracked down. If anyone can provide steps to repro or get a log (see https://github.com/dart-code-checker/dart-code-metrics/issues/262#issuecomment-819699313) that would definitely help understand what's going on. Thanks!

**Comment by makstheimba** [comment]
> 2. I've encountered with a different issue in VS Code (when LSP preview is enabled), but I think that they have the same problem underneath - if the `plugins` entry is added to the `analysis_options`, then the analyser starts working like that
> 
>  Screen.Recording.2021-04-10.at.12.23.48.mov 
> Or in other words - each time the file is updated and not saved, error disappear. And if I remove `plugins` and restart the analysis server then everything works fine and without this glowing. And again, I used `3.0.0-nullsafety.1` where plugin does nothing.

@incendial 
Hi! Any updates on this? We experience the same flickering behavior in VS Code when using dart-code-metrics 4.1.0 plugin with analyzer

**Comment by merelj** [comment]
same issue, my dart version is:
Dart SDK version: 2.13.3 (stable) (Wed Jun 9 12:44:44 2021 +0200) on "macos_x64"

**Comment by makstheimba** [comment]
@incendial Thanks for noticing my comment 🙏 

My Dart version:
`Dart SDK version: 2.13.4 (stable) (Wed Jun 23 13:08:41 2021 +0200) on "macos_x64"`

Flutter doctor:
```
[✓] Flutter (Channel stable, 2.2.3, on Mac OS X 10.15.7 19H15 darwin-x64, locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
[✓] Xcode - develop for iOS and macOS
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.1)
[✓] VS Code (version 1.59.1)
[✓] Connected device (1 available)

• No issues found!
```

**Comment by incendial** [comment]
@merelj @makstheimba if I'm not mistaken, flickering issues with enabled LSP mode were fixes in 2.14.

**Comment by incendial** [comment]
Closing due to missing reproduction steps. Feel free to reopen if the issue is still there and has repro steps in the VS Code.
